### DISCUSSION TITLE
Controlled mutability managed references

### DISCUSSION THREAD

**Discussion by Sergio0694** [dotnet/csharplang#5590] [Language Ideas]
## Overview

Today, the way `ref`-s work in C# is "all or nothing": you either have `ref` which gives consumers complete control over the target memory location, or `ref readonly`, which makes consumers completely unable to make any kind of modifications on the target memory location, even indirectly (eg. invoking a non readonly instance member on the target value results in a shadow copy). This is not ideal, as there are many scenarios where library authors might want to give consumers the ability to still invoke any members on the target value, but with the restriction that that value cannot be _assigned_.

This is a concept that C# lacks today, but that has always existed in IL: controlled mutability pointers.

> **III.1.8.1.2.2**: "The `readonly.` prefix and `unbox` instructions can produce what is called a _controlled-mutability_ 
managed pointer. Unlike ordinary managed pointer types, a _controlled-mutability_ managed 
pointer is not verifier-assignable-to."

In practice, these pointers are just like `ref`, with the main difference that they cannot be assigned to. This would be the perfect sweet spot between `ref` and `in` when authors want to offer more control over a value that must not be reassigned.

## Some use case examples

- If a `struct` type is used in a `using` block/declaration, it cannot be passed by `ref`, as callers would be able to reassign to it. Only being able to pass it via `in` though is extremely limiting though, as callers will just not be able to invoke any instance members on it that would mutate it in any way. This is completely different than how it works when the type in a `using` block/declaration is a reference type: here you still cannot reassign to it, but you're free to mutate it through instance members. Being able to pass a `struct` in a `using` via `ref const` would bring feature parity with reference types here.
- APIs that expose results of unbox operations (through `Unsafe.Unbox<T>`) would be able to return the value as a `ref const`, which would make consumers unable to break the spec by accidentally reassigning to it (which is illegal and could break).
- Any other scenarios where consumers might want to return a reference to eg. some struct member allowing callers to invoke public methods on it that might change its state, but without them being able to reassign that whole instance.

## Proposal

Without any new keywords, the `ref const` (or alternatively, `const ref`, whichever is deemed best) could be used to indicate a _controlled-mutability_ managed pointer. It would behave exactly like a `ref`, but assigning to it would result in a build error.

- Converting a `ref` to a `ref const` is legal.
- Converting a `ref const` to an `in`/`ref readonly` is legal.
- Converting a `ref const` to `ref` or `out` is illegal.

Additionally, some notes:

- Passing value in a `using` block/declaration via `ref const` is legal.
- Passing a `readonly` field via `ref const` is illegal (as it could be mutated).

## Grammar

```antlr
argument
    : (identifier ':')? refout=(REF | REF CONST | OUT | IN)? (VAR | type_)? expression
    ;

explicit_anonymous_function_parameter
    : refout=(REF | REF CONST | OUT | IN)? type_ identifier
    ;

local_variable_declaration
    : (USING | REF | REF CONST | REF READONLY)? local_variable_type local_variable_declarator ( ','  local_variable_declarator { this.IsLocalVariableDeclaration() }? )*
    // Existing grammar
    ;

typed_member_declaration
    : (REF | REF CONST | READONLY REF | REF READONLY)? type_
    // Existing grammar
    ;

parameter_modifier
    : REF CONST
    | REF CONST THIS
    // Existing grammar
    ;

interface_member_declaration
    : attributes? NEW?
        (UNSAFE? (REF | REF CONST | REF READONLY | READONLY REF)? type_
        ( identifier type_parameter_list? OPEN_PARENS formal_parameter_list? CLOSE_PARENS type_parameter_constraints_clauses? ';'
    // Existing grammar
    ;
```

## Metadata encoding

Much like with `in`, this could be implemented by `ref const` using an attribute.
For instance, `[System.Runtime.CompilerServices.IsControlledMutabilityAttribute]`, which would support this feature.

- For `ref const` params, they would be `ref [IsControlledMutability]`.
- For `ref const` returns, they would be `ref modreq([IsControlledMutability])`.

**Comment by HaloFour** [comment]
> This is not ideal, as there are many scenarios where library authors might want to give consumers the ability to still invoke any members on the target value, but with the restriction that that value cannot be assigned.

The compiler shallow-copies the reference as it has no idea if the method would cause a reassignment (unless the member has been marked as `readonly`).  How do you propose to solve that issue here?

**Reply by Sergio0694** [reply]
The shallow-copy would not be performed when the value is passed as a controlled-mutability reference. The compiler would emit code that is equivalent to what you would get today in C# by doing:

```csharp
using Foo foo = default;

SomeMethod(ref Unsafe.AsRef(in foo));
```

Where `SomeMethod` would be:

```csharp
void SomeMethod(ref const Foo foo)
{
}
```

The compiler will see that the `foo` parameter in `SomeMethod` is a `ref const` (eg. the `ref` could be encoded with a `[System.Runtime.CompilerServices.IsControlledMutabilityAttribute]`, much like `in`), and it would allow callers to pass a non-assignable-to value via `ref`. That is, callers could just do:

```csharp
using Foo foo = default;

SomeMethod(ref foo); // This works, because the param is ref const
```

This would all work fine because `SomeMethod` will never allow assignments to `foo`.

**Reply by HaloFour** [reply]
How is that different from `in`?  That enables the same scenario and prevents reassignment.

**Reply by Sergio0694** [reply]
The key difference is that `in` doesn't allow _any_ modifications on the parameter. If `SomeMethod` invoked an instance member on `foo` that was not readonly, the compiler would introduce a shallow copy and invoke the method on it. With a `ref const` instead, you cannot _reassign_ the parameter, but you can still invoke mutating members on it, so you can do side effect on them. This is critical in all those scenarios I've mentioned in the proposal, and it's something that's not doable at all today without using unsafe hacks, losing a lot of expressiveness and making the code way more error prone.

Consider this:

```csharp
struct Foo foo
{
    int x;

    public void Increment() => x++;
}

SomeMethod(in Foo foo)
{
    foo.Increment(); // Trap! 💣
}

SomeMethod(ref const Foo foo)
{
    foo.Increment(); // This is fine!

    foo = default; // Build error
}
```

**Reply by HaloFour** [reply]
So you do want to allow reassignment of the struct, just not within the method declared as `ref const`.  IMO that's a much bigger trap.

IMO that kind of functionality would be enabled by `readonly` parameters: https://github.com/dotnet/csharplang/issues/188

`readonly ref` vs. `ref readonly`

Although honestly I'm not sure that would support such behavior either as the expectation is that something `readonly` cannot be reassigned, and mutation a struct is effectively reassignment.  The defensive copy strategy is also taken with `readonly` fields, for example.

**Reply by Sergio0694** [reply]
> "So you do want to allow reassignment of the struct"

No, I do not. I want to allow _mutating_ the struct, but only in a manner that depends on the public API surface of the specific types. Some types might eg. only expose an API to mutate some of their internal state, and this would allow consumers to only do that, without just being able to reassign the whole value and breaking everything. Consider the example of the `using` value: if consumers were able to reassign the whole value, you could cause memory leaks or worse, if that value was eg. holding resources that needed to be disposed. With `ref const` instead, consumers would only be able to use public APIs to mutate that value in a controlled way, setup by the author of those types. It's a critical difference. You can call accessible mutable members, but cannot _reassign_ the whole instance.

> "that kind of functionality would be enabled by `readonly` parameters"

I'm not sure it would, especially not in all cases. You can easily see how eg. this scenario with the `using` value I described above would completely break down with `readonly` parameters. By that I mean, it would just not work at all:

```csharp
using Foo foo = default;

SomeMethod(ref foo); // Build error: can't pass a using value by `ref`

void SomeMethod(readonly ref Foo foo)
{
}
```

Furthermore, `readonly` params is, well, for parameters. You cannot return them.
With `ref const` instead, you can have members returning a `ref const`, just like `ref readonly`.

**Reply by HaloFour** [reply]
I see.  I guess I can see the utility to that, although I don't think the term `ref const` accurately reflects that intent.

I believe that this came up during the design of `ref readonly` and is partially why the keywords are specifically in that order.  See this comment by a previous member of the team: https://github.com/dotnet/csharplang/issues/38#issuecomment-279844069

I don't believe that was expanded upon in design meetings.

**Reply by Sergio0694** [reply]
> "I don't think the term `ref const` accurately reflects that intent."

Sure, that wass mostly just the closest I could come up with by reusing existing keywords. I do think `const` would work given it's also somewhat similar to how it is in C++, plus as you said too, the modifier has to be _after_ `ref`. And we definitely wouldn't want a new keyword just for this, as it's admittedly a niche feature. Would you have other suggestions other than `ref const`?

**Reply by CyrusNajmabadi** [reply]
I agree that this would be subsumed by the `ref readonly` (or `readonly ref`) work.

**Reply by Sergio0694** [reply]
I don't see how the two would be exactly the same but hey, as long as I can functionally use the same way I'm not complaining. My main issue though is: will it be allowed to pass a value in a `using` by `readonly ref`? Because that's one of the main motivations for this feature. I mean does the `readonly ref` proposal also include an update to the spec to make that legal?
Also, isn't that proposal only for parameters? Or does it include `readonly ref` returns too?

**Reply by CyrusNajmabadi** [reply]
> I don't see how the two would be exactly the same

We support `ref readonly` today.  A reference to something readonly.  A `readonly ref` would be where the reference itself can only be read, not written to.

**Reply by Sergio0694** [reply]
> "A `readonly ref` would be where the reference itself can only be read, not written to."

I feel like that phrasing is a bit confusing, does that mean you can just not do `x = Something()`, but calling any (mutating) member on `x` via that `ref` is fair game? Would passing a `using` value via `readonly ref` be allowed? Would `readonly ref` be allowed for returns? And also, once we get `ref` fields, how would you indicate a `readonly ref` field? According to Jared's proposal, all `ref` fields need to be `readonly`, so that'd make it... `readonly readonly ref`? 🤔

**Reply by jaredpar** [reply]
> A readonly ref would be where the reference itself can only be read, not written to.

My vision for `readonly ref` was slightly different. It's a `ref` which is fully mutable, that is it can be assigned, methods can be called on it, but it cannot be `ref` assigned. Essentially consider the following code today: 

``` c#
int local1 = 42;
int local2 = 13;
ref int r1 = ref local1;
r1 = ref local2;
Console.WriteLine(r1); // Prints 13
```

The vision I'd had for `readonly ref` was it's a `ref` that can't be re-assigned so ...

``` c#
int local1 = 42;
int local2 = 13;
readonly ref int r1 = ref local1;
r1 = 100; // Okay no prob here
r1 = ref local2; // Error: can't ref assign a readonly ref 
Console.WriteLine(r1); 
```

This state of `ref` is essentially what you have with `ref` fields. The ability to `ref` assign is **very** limited because it's hard to account for in the safety rules.

**Reply by Sergio0694** [reply]
Right, yeah this makes perfect sense to me and it's what I was expecting when I first read the `readonly` proposal as well: the `readonly` modifier refers to the parameter, so just like it makes a value readonly (eg. `readonly int`), it would make the _ref_ readonly if it's a `readonly ref int`, but it would give you no change in semantics on the target memory location. Which leads to my next question then: isn't this a separate feature than controlled mutability then? After talking with @jnm2, we realized that `using`-s on a value type essentially already behave exactly this way: you can call mutating members on that value, but you cannot assign to it. There is just no way to "pass this semantics along" by ref. You can either only pass by `ref`, which allows consumers to do whatever, or you can pass by `in`, which allows consumers to not mutate any state at all. You can only pass a value in a `using` by `in`, which means you've now lost the ability to mutate it which was allowed instead, if you just referenced that local directly. Do you reckon it would make sense to add this "intermediate" mutability type? 🤔

Note: regarding the `using` example I talked about with Cyrus, yes this wouldn't be enough. One would also need the ability to declare a type as `unique` for it to work correctly (ie. a `unique ref const T` reference would only allow you to call mutating members on it, but you wouldn't be able to make copies nor to move it out, nor to move another value into it).

**Comment by sharwell** [comment]
It looks like the proposal is indistinguishable from a separate proposal to simply allow a `using` variable to be passed by `ref` if it is a value type. Reassignment of the parameter would be the same as calling a mutating member on the parameter that reassigns `this`.

**Reply by Sergio0694** [reply]
Wouldn't that potentially lead to all sorts of issues though? If a value type in a `using` is holding unmanaged resources, passing it via `ref` would allow consumers to (accidentally) reassign it and break things, or cause memory leaks. The difference with an instance method reassigning `this` is that that instance method would be defined in the actual type, so the author of that type, not the consumer, would have control over what can be reassigned or not. It doesn't seem to me like just allowing `using` value types over `ref` would be as safe as this proposal, from what I can see 🤔
It is the same as with a reference type in a `using`: you can only mutate it via its public methods, which mean the author of that type and not the consumers have full control over what is assigned. A consumer can't just reassign (ie. swap in this case, if it's a reference type) the object in a `using` and break things. This would be the case with `ref` for `using` instead.
Not to mention, this proposal would also allow `ref const` returns, for instance.

**Reply by Sergio0694** [reply]
Also forgot to mention, the proposal also called out the fact that this would also allow expressing APIs such as `Unsafe.Unbox<T>`, which specifically call out the return (from `unbox.any`) being a "controlled mutability managed pointer", which is otherwise not expressible in C#. Just allowing value types in a `using` to be passed by `ref` wouldn't solve that, and also it'd be much more dangerous for the reason mentioned above (it'd take away control from the authors of the types being used in a `using` about how to allow consumers to mutate those instances).

**Reply by john-h-k** [reply]
> Reassignment of the parameter would be the same as calling a mutating member on the parameter that reassigns this.

Yes but design-wise that is completely different, as that is the object mutating itself rather than being arbitrary reassigned or defaulted by the consumer

**Reply by Sergio0694** [reply]
+1 to this. The distinction here is crucial and the whole point of this proposal. While a self-reassignment is functionally equivalent, being able to restrict the ability of doing that to _type authors_ only, and not consumers, is really important, and where I think the value of this proposal lies.
